src/garrysmod/gamemodes/zombiesurvival/gamemode/howl/sv_command.lua

```diff
@@ -0,0 +1,95 @@
+local Command = {}
+Command.prototype = {}
+Command.__index = Command.prototype
+
+Command.argument_types = {}
+
+setmetatable(Command, {
+    __call = function(self, name, perm)
+        local obj = {}
+        obj.name = name or ""
+        obj.perm = perm or nil
+        obj.argument_types = {}
+        obj.required_args = #obj.argument_types
+        
+        return setmetatable(obj, self)
+    end
+})
+
+function Command.prototype:register_argument(argument_name, argument, is_required)
+    assert(Command.argument_types[argument] ~= nil)
+    
+    table.insert(self.argument_types, { name = argument_name, type = argument, is_required = is_required })
+    if is_required then
+        self.required_args = self.required_args + 1
+    end
+end
+
+function Command.prototype:usage()
+    local msg = string.format("usage: %s", self.name)
+    
+    for _, arg in ipairs(self.argument_types) do
+        if arg.is_required then
+            msg = string.format("%s <%s>", msg, arg.name)
+        else
+            msg = string.format("%s [%s]", msg, arg.name)
+        end
+    end
+    
+    return msg
+end
+
+function Command.prototype:parse(raw_args)
+    if #raw_args < self.required_args then
+        return string.format(
+            "%s\nNot enough arguments. %d expected. %d provided.",
+            self:usage(),
+            self.required_args,
+            #raw_args,
+        )
+    end
+
+    local parsed_args = {}
+
+    for i = 1, #self.argument_types do
+        local arg
```

Comment:
We're iterating over the arguments we're saying this `Command` instance takes, rather than the amount of arguments the command received. This way, if there are excess arguments in the `raw_args` passed to the command, we can roll them up into a single value to try and parse. 

This is useful for something like `!ban`, where the last argument is the ban reason, and we'd just want to concatenate every single word at the end of the command into a single string.